File: LeetCode/76. 最小覆盖子串.cpp
class Solution
{
public:
  string minWindow(string s, string t)
  {
    if (s.length() == 0 || t.length() == 0 || s.length() < t.length())
      return "";
    unordered_map<char, int> tFreq;
    for (char c : t)
      tFreq[c]++;
    const int sLen = s.length(), tLen = t.length();
    int begin = 0, l = 0, r = 0, minLen = sLen + 1, d = tLen;
    while (r < sLen)
    {
      const char rc = s[r];
      if (!tFreq.count(rc))
      {
        r++;
        continue;
      }
      if (tFreq[rc] > 0)
      {
        d--;
      }
      tFreq[rc]--;
      r++;
      while (d == 0)
      {
        const char lc = s[l];
        if (r - l < minLen)
        {
          minLen = r - l;
          begin = l;
        }
        if (!tFreq.count(lc))
        {
          l++;
          continue;
        }
        if (tFreq[lc] == 0)
        {
          d++;
        }
        tFreq[lc]++;
        l++;
      }
    }
    return minLen == sLen + 1 ? "" : s.substr(begin, minLen);
  }
};

File: LeetCode/76. 最小覆盖子串.ts
var minWindow = function(s: string, t: string): string {
  let tFreq=new Map();
  for(let c of t){
      tFreq.set(c,tFreq.has(c)?tFreq.get(c)+1:1);
  }
  let begin=0,minLen=s.length+1;
  let l=0,r=0;
  let d=t.length;
  while(r<s.length){
      let rc=s[r];
      if(!tFreq.has(rc)){
          r++;
          continue;
      }
      if(tFreq.get(rc)>0){
          d--;
      }
      tFreq.set(rc,tFreq.get(rc)-1);
      r++;
      while(d===0){
          let lc=s[l];
          if(!tFreq.has(lc)){
              l++;
              continue;
          }
          if(r-l<minLen){
              minLen=r-l;
              begin=l;
          }
          if(tFreq.get(lc)===0){
              d++;
          }
          tFreq.set(lc,tFreq.get(lc)+1);
          l++;
      }
  }
  return (minLen===s.length+1)?"":s.substr(begin,minLen);
};